zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ExporterDirectorPartitionTransitionStep.java

```diff
@@ -122,8 +144,44 @@ private ActorFuture<Void> openExporter(
                 director.resumeExporting();
                 break;
             }
+
+            // The config might have changed after ExporterDirector has created
+            disableExportersIfConfigChanged(exporterDescriptors, context);
           }
         });
     return startFuture;
   }
+
+  private void disableExportersIfConfigChanged(
+      final Collection<ExporterDescriptor> startedExporters,
+      final PartitionTransitionContext context) {
+    final var currentEnabledExporters = getEnabledExporterDescriptors(context);
+
+    for (final var exporter : startedExporters) {
+      if (!currentEnabledExporters.contains(exporter)) {
+        context.getExporterDirector().disableExporter(exporter.getId());
+      }
+    }
+    // TODO: Enable newly enabled exporters when the functionality is added
+  }
+
+  private static Collection<ExporterDescriptor> getEnabledExporterDescriptors(
+      final PartitionTransitionContext context) {
+    final Collection<ExporterDescriptor> exporterDescriptors = context.getExportedDescriptors();
+    final var exporterConfig = context.getDynamicPartitionConfig().exporting().exporters();
+    return exporterDescriptors.stream()
+        .filter(exporterDescriptor -> isEnabled(exporterConfig, exporterDescriptor))
+        .toList();
+  }
+
+  private static boolean isEnabled(
+      final Map<String, ExporterState> exporterConfig,
+      final ExporterDescriptor exporterDescriptor) {
+    // TODO: If the exporter is not found in exporterConfig, it should be considered as disabled.
+    // But we can do that only after https://github.com/camunda/zeebe/issues/18296 is implemented.
+    // Until then, we assume if the exporter is not found in the config, it is considered as
+    // enabled. Note that this is a temporary workaround.
+    return !exporterConfig.containsKey(exporterDescriptor.getId())
+        || exporterConfig.get(exporterDescriptor.getId()).state() == State.ENABLED;
```

Comment:
Unfortunately this temporary hack is required because the update tests and another test is failing. This should be fixed after https://github.com/camunda/zeebe/issues/18296 this is implemented.

This also raises another question. What should be the behavior when a broker restarts with a new configured exporter? Should the export starts as disabled or enabled? My preference is - it starts as disabled since it is not in the dynamic config and an explicit "enable" request can be used to enable this. We have one test failing because it starts zeebe with no exporters and then restart with a recording exporter and uses the recording exporter to verify the state. With the above temp hack it work now, but this will change.

Reply:
Let's discuss it f2f.